man/man1/play_genre_artist.1
.\" Automatically generated by Pandoc 2.19.2
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "PLAY_GENRE_ARTIST" "1" "February 13, 2022" "play_genre_artist 2.0.1" "User Manual"
.hy
.SH NAME
.PP
play_genre_artist - Play Roon Library Artist in Genre
.SH SYNOPSIS
.PP
\f[B]play_genre_artist\f[R] [ -z ZONE ] [ GENRE ] [ ARTIST ] [ EXGENRE ]
[ EXARTIST ]
.SH DESCRIPTION
.PP
Plays Roon Library genre\[aq]s artist matching the specified artist name
or substring.
This command is invoked by \[dq]roon -g genre [-a artist] [-x exgenre]
[-X exartist]\[dq].
.SH COMMAND LINE OPTIONS
.TP
\f[B]-z ZONE\f[R]
Specify the Roon Zone in which to play.
If no zone is specified the default zone will be used.
\[aq]-z default\[aq] indicates the default zone, \[aq]-z last\[aq] the
last zone used
.TP
\f[B]GENRE\f[R]
Provide a genre name, in quotes if it contains spaces, or a
case-sensitive substring.
All Roon library genre names matching the provided genre name or
containing the substring will be searched for artists
.TP
\f[B]ARTIST\f[R]
Provide an artist name, in quotes if it contains spaces, or a
case-sensitive substring.
All Roon library artist names matching the provided artist name or
containing the substring will be returned and the first one returned
will be played.
.TP
\f[B]EXGENRE\f[R]
Provide a genre exclusion string, in quotes if it contains spaces.
No Roon library genre names including the provided genre exclusion
string will be searched
.TP
\f[B]EXARTIST\f[R]
Provide an artist exclusion string, in quotes if it contains spaces.
No Roon library artist names including the provided artist exclusion
string will be played
.SH EXAMPLES
.TP
\f[B]play_genre_artist Prog Yes\f[R]
Will play the first artist returned in a list of all Roon library artist
names with names containing the string \[dq]Yes\[dq] by the genres whose
name contains \[dq]Prog\[dq]
.TP
\f[B]play_genre_artist Rock Johnny foo Cash\f[R]
Will play the first artist returned in a list of all Roon library artist
names with names containing the string \[dq]Johnny\[dq] but not
\[dq]Cash\[dq] by the genres whose name contains \[dq]Rock\[dq] but not
\[dq]foo\[dq]
.SH AUTHORS
.PP
Written by Ronald Record <github@ronrecord.com>
.SH LICENSING
.PP
PLAY_GENRE_ARTIST is distributed under an Open Source license.
See the file LICENSE in the PLAY_GENRE_ARTIST source distribution for
information on terms & conditions for accessing and otherwise using
PLAY_GENRE_ARTIST and for a DISCLAIMER OF ALL WARRANTIES.
.SH BUGS
.PP
Submit bug reports online at:
<https://github.com/doctorfree/RoonCommandLine/issues>
.SH SEE ALSO
.PP
\f[B]get_core_ip\f[R](1), \f[B]get_zone_info\f[R](1),
\f[B]get_zones\f[R](1), \f[B]list_albums\f[R](1),
\f[B]list_artist_albums\f[R](1), \f[B]list_artists\f[R](1),
\f[B]list_composers\f[R](1), \f[B]list_genre_albums\f[R](1),
\f[B]list_genre_artists\f[R](1), \f[B]list_genres\f[R](1),
\f[B]list_playlists\f[R](1), \f[B]list_radio\f[R](1),
\f[B]list_tags\f[R](1), \f[B]list_zones\f[R](1),
\f[B]play_album\f[R](1), \f[B]play_artist\f[R](1),
\f[B]play_artist_album\f[R](1), \f[B]play_artist_track\f[R](1),
\f[B]play_composer\f[R](1), \f[B]play_genre\f[R](1),
\f[B]play_genre_album\f[R](1), \f[B]play_genre_artist\f[R](1),
\f[B]play_playlist\f[R](1), \f[B]play_radio\f[R](1),
\f[B]play_tag\f[R](1), \f[B]roon\f[R](1), \f[B]set_zone\f[R](1),
\f[B]set_zone_group\f[R](1), \f[B]zone_command\f[R](1)
.PP
Full documentation and sources at:
<https://github.com/doctorfree/RoonCommandLine>
.PP
The corresponding Python script,
\f[I]/usr/local/Roon/api/play_genre_artist.py\f[R], can be invoked
directly as:
.PP
\f[B]python3 /usr/local/Roon/api/play_genre_artist.py [ options]\f[R]
.PP
Where [ options ] are described in the following usage message:
.IP
.nf
\f[C]
usage: play_genre_artist.py [-h] [-a ARTIST] [-g GENRE] [-X EXARTIST]
                            [-x EXGENRE] [-z ZONE]

optional arguments:
  -h, --help            show this help message and exit
  -a ARTIST, --artist ARTIST
                        artist search term
  -g GENRE, --genre GENRE
                        genre search term
  -X EXARTIST, --exartist EXARTIST
                        artist exclude search term
  -x EXGENRE, --exgenre EXGENRE
                        genre exclude search term
  -z ZONE, --zone ZONE  zone selection
\f[R]
.fi


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

May 30, 2024 :
RoonCommandLine version 2.1.3 release 1

    This release provides support for `-z default` and `-z last` to specify default zone and last zone used
        * Add usage to `get_zone_info`
        * Updated man pages
        * Add `-P` flag to `now_playing` to list paused zones
        * Upgrade checks to prevent malformed `roon_api.ini`

May 27, 2024 :
RoonCommandLine version 2.1.2 release 4

    This release provides improved info and formatting for now playing and media search
        * `now_playing` info includes state
        * `now_playing` output formatted with `rich` if available
        * Fixes for upgrades on systems with BSD sed (e.g. macOS)
        * Media search and selection enhancements

May 23, 2024 :
RoonCommandLine version 2.1.2 release 3

    This release provides support for signal handling and fuzzy track search in menus
        * Add signal handler to Python scripts
        * Add tracks listing, track play, and album track play commands
        * Add Examples and Usage to main menu
        * Improve info when selecting and listing in menus
        * Install `jq` in `macInstall`

May 20, 2024 :
RoonCommandLine version 2.1.2 release 2

    This release provides support for installation and use of `roon-tui`
        * Install `roon-tui` with `roon -I`
        * Add `roon-tui` wrapper `roontui`
        * Add shortcuts to menus
        * Add `list_album_tracks` command
        * Use `fzf` to select media and list media

May 16, 2024 :
RoonCommandLine version 2.1.1 release 1

    This release provides support for shuffled artist/genre/playlist playback
        * Shuffled playback is accomplished using the play_media `Shuffle` action
        * All play commands now accept a `-z ZONE` argument to specify playback zone

May 09, 2024 :
RoonCommandLine version 2.1.0 release 6

    This release provides a more robust Python virtual environment installation of pip:
        * If ensurepip Python module not available then download and run get-pip.py
        * Requires curl or wget

May 08, 2024 :
RoonCommandLine version 2.1.0 release 5

    This release ensures that pip is installed in the Python virtual environment:
        * Activate the Python virtual environment in all install scripts
        * Use the built-in ensurepip Python module to ensure pip is installed
        * Fix timing after first discovery to avoid requiring dual authorization

Feb 20, 2024 :
RoonCommandLine version 2.1.0 release 4

    This release adds the capability to transfer the current queue to another zone:
        * Add `transfer_zone` command and Python backend
        * Stop core discovery after core discovered

Feb 14, 2024 :
RoonCommandLine version 2.1.0 release 3

    This release provides support for Ubuntu 23.04, Debian 12, and later:
        * Use a Python virtual environment
        * Additional checks for valid saved roon_api.ini
        * Add support for `q` to quit
        * Migrate to Github
        * Add lolcat colored figlet text transparent PNG header

May 16, 2023 :
RoonCommandLine version 2.0.8 release 5

    This is a bug fix release:
        * PEP 8 style guide conformance
        * Improved interactive menu system
        * Fix shellcheck warnings
        * Fix bug in call to get_zones in Get/Set Zone

May 2, 2023 :
RoonCommandLine version 2.0.8 release 4

    This is a bug fix release:
        * Only restore saved roon_api.ini if in new format

April 29, 2023 :
RoonCommandLine version 2.0.8 release 3

    This release adds support for:
        * Check for python3 in roon set_default
        * Support for Homebrew installed Python etc
        * Don't error out in postinstall
        * Postinstall fails when there is a directory that is not a valid user
        * Check if run as root user in Install
        * Always upgrade roonapi during install
        * Mac installation fixes
        * Restore ensurepip in Mac install
        * Fix permissions errors during Mac install
        * Remove symlinks in /usr/local/bin in Mac uninstall
        * Update install size in Debian packaging
        * Redirect stderr when invoking tput to set BOLD and NORMAL text
        * In Linux and Mac manual install scripts, upgrade roonapi if installed
        * Update set_volume man page, changelog, and release notes
        * Support percentage volume change using new roonapi methods
        * Handle negative volume levels in roon_fade

January 19, 2023 :
RoonCommandLine version 2.0.8 release 2 includes Mac installation fixes

December 31, 2022 :
RoonCommandLine version 2.0.8 release 1 provides API compatibility and feature updates

    This release adds support for:
        * Compatibility with Python Roon API 0.1.3
        * Support for percentage volume change/setting
        * Add immediate mode to roon_fade with `now` argument
        * Implement non-interactive immediate fading
        * Add interactive menu system (`roon_fade -i`)
        * Set man permissions correctly in manual install scripts

December 11, 2022 :
RoonCommandLine version 2.0.7 release 1 is a minor feature and bugfix update

    This release adds support for:
        * Save authentication token right after we get a RoonApi handle (bug fix)
        * Add options to `roon` command for controlling fading
        * Add fade feature (`roon_fade` user command and `roon_faded` fade daemon)
        * Man pages for fading and new commands
        * Fixes for `now_playing`, `get_zone_volume`, and `get_zone_attributes`
        * Add `get_zone_remaining` to get remaining time for now playing in zone
        * Add `get_zone_volume` and `get_zone_attributes.py`

November 15, 2022 :
RoonCommandLine version 2.0.6 release 2 is a minor feature update

    This release adds support for:
        * Add list_playlist_tracks to list a playlist's tracks
        * Run roonapi commands in postinstall as user if provided

api/play_genre_artist.py
import argparse
import configparser
from os import path
import signal
import sys
from roonapi import RoonApi

def signal_handler(sig, frame):
    sys.exit(0)
signal.signal(signal.SIGINT, signal_handler)

config = configparser.ConfigParser()
config.read('/usr/local/Roon/etc/roon_api.ini')

# Set to IP address of your Roon Core
server = config['DEFAULT']['RoonCoreIP']
# Set to Port of your Roon Core
port = config['DEFAULT']['RoonCorePort']
# Name of the file that holds a Roon API token
tokenfile = config['DEFAULT']['TokenFileName']

parser = argparse.ArgumentParser()
parser.add_argument("-a", "--artist", help="artist search term")
parser.add_argument("-g", "--genre", help="genre search term")
parser.add_argument("-X", "--exartist", help="artist exclude search term")
parser.add_argument("-x", "--exgenre", help="genre exclude search term")
parser.add_argument("-z", "--zone", help="zone selection")
args = parser.parse_args()

if args.artist:
    artistsearch = args.artist
else:
    artistsearch = config['DEFAULT']['DefaultArtist']
if args.genre:
    genresearch = args.genre
else:
    genresearch = config['DEFAULT']['DefaultGenre']
if args.exartist:
    exartistsearch = args.exartist
else:
    exartistsearch = None
if args.exgenre:
    exgenresearch = args.exgenre
else:
    exgenresearch = None
if args.zone:
    target_zone = args.zone
else:
    target_zone = config['DEFAULT']['DefaultZone']

version = config['DEFAULT']['RoonCommandLineVersion']
release = config['DEFAULT']['RoonCommandLineRelease']
fullver = version + "-" + release

appinfo = {
    "extension_id": "roon_command_line",
    "display_name": "Python library for Roon",
    "display_version": fullver,
    "publisher": "RoonCommandLine",
    "email": "roon@ronrecord.com",
    "website": "https://github.com/doctorfree/RoonCommandLine",
}

# Can be None if you don't yet have a token
if path.exists(tokenfile):
    token = open(tokenfile).read()
else:
    token = "None"

roonapi = RoonApi(appinfo, token, server, port)

# save the token for next time
with open(tokenfile, "w") as f:
    f.write(str(roonapi.token))

# get target zone output_id
outputs = roonapi.outputs

output_id = None
for (k, v) in outputs.items():
    if target_zone in v["display_name"]:
        output_id = k

if output_id is None:
    err = "No zone found matching " + target_zone
    sys.exit(err)
else:
    # List matching genres
    genres = roonapi.list_media(output_id, ["Genres", genresearch])
    if genres:
        artist = None
        for genre in genres:
            if exgenresearch is not None:
                if exgenresearch in genre:
                    continue
            # List matching artists
            artists = roonapi.list_media(output_id,
                                         ["Genres", genre,
                                          "Artists", artistsearch])
            if exartistsearch is not None and artists:
                artists = [chk for chk in artists if exartistsearch not in chk]
                if artists:
                    artist = artists[0]
                    print("Playing artist name", artist, "in", genre, "genre")
                    roonapi.play_media(output_id,
                                       ["Genres", genre,
                                        "Artists", artist], None, False)
                    if len(artists) > 1:
                        print("\nArtist names in", genre,
                              "genre matching", artistsearch, ":\n")
                        print(*artists, sep="\n")
                        print("\nTo play another artist in this genre by name either specify")
                        print("the full name or enough of a substring to provide a single match\n")
                    break
        if artist is None:
            print("No artists found matching", artistsearch)
    else:
        print("No genres found matching ", genresearch)
